crypto/ube/ube.c

```diff
@@ -0,0 +1,238 @@
+      ret_snapsafe_gn == 0) {
+    return 0;
+  }
+
+  return 1;
+}
+
+// ube_is_detected computes whether a UBE has been detected or not by comparing
+// the cached detection generation numbers with the current detection generation
+// numbers. The current generation numbers must be loaded into
+// |current_detection_gn| before calling this function.
+//
+// Returns 1 if UBE has been detected and 0 if no UBE has been detected.
+static int ube_is_detected(struct detection_gn *current_detection_gn) {
+
+  if (ube_state.cached_fork_gn != current_detection_gn->current_fork_gn ||
+      ube_state.cached_snapsafe_gn != current_detection_gn->current_snapsafe_gn) {
+    return 1;
+  }
+  return 0;
+}
+
+// stays invariant after a failure state has been reached.
+int get_ube_generation_number(uint64_t *current_generation_number) {
+
+  GUARD_PTR(current_generation_number);
+  *current_generation_number = 0;
+
+  CRYPTO_once(&ube_state_initialize_once, ube_state_initialize);
+
+  // If something failed at an earlier point short-circuit immediately. This
+  // saves work in case the UBE detection is not supported. The check below
+  // must be done after attempting to initialize the UBE state. Because
+  // initialization might fail and we can short-circuit here.
+  if (ube_detection_unavailable == 1) {
+    return 0;
+  }
+
+  struct detection_gn current_detection_gn = { 0 };
+
+  // First read generation numbers for each supported detection method. We do
+  // not mutate |ube_state|. So, a read lock is sufficient at this point. Each
+  // individual detection method will have their own concurrency controls if
+  // needed.
+
+  CRYPTO_STATIC_MUTEX_lock_read(&ube_lock);
+  if (ube_get_detection_generation_numbers(&current_detection_gn) != 1) {
+    ube_failed();
+    CRYPTO_STATIC_MUTEX_unlock_read(&ube_lock);
+    return 0;
+  }
+  if (ube_is_detected(&current_detection_gn) == 0) {
+    // No UBE detected, so just grab UBE generation number from the state.
+    *current_generation_number = ube_state.generation_number;
+    CRYPTO_STATIC_MUTEX_unlock_read(&ube_lock);
+    return 1;
+  }
+  CRYPTO_STATIC_MUTEX_unlock_read(&ube_lock);
+
+  // Reaching this point means that an UBE has been detected. We must now
+  // synchronize an update to the UBE generation number. To avoid redundant
+  // reseeds, we must ensure the generation number is only incremented once for
+  // all UBE's that might have happened. Therefore, first take a write lock but
+  // before mutation the state, check for an UBE again. Checking again ensures
+  // that only one thread increments the UBE generation number, because the
+  // cached detection method generation numbers have been updated by the thread
+  // that had the first entry.
+
+  CRYPTO_STATIC_MUTEX_lock_write(&ube_lock);
+  if (ube_get_detection_generation_numbers(&current_detection_gn) != 1) {
+    ube_failed();
+    CRYPTO_STATIC_MUTEX_unlock_write(&ube_lock);
+    return 0;
+  }
```

Comment:
Instead of having to lock/unlock here did you consider implementing the counters in `ube_state` using atomic counters? Just thinking about our faster new c11 atomics and all the cool places we can use them instead of pthreads locks.

Reply:
I'm not sure we can replace synchronising here with atomic operations only. Because we also synchronise to ensure that we only increment once per UBE. Is there another primitive that ensures that?

If we do not mutex this entire block, how can we ensure that all logic in `ube_is_detected` use the updated global state?

Reply:
I was thinking we could make the global counters atomics, but that wouldn't get us the property that they are only incremented once per UBE, two threads could increment it at the same time and that might cause additional seeding which would be inefficient.